Algebraic-Diagrammatic-Construction/adc_helper.py

```diff
@@ -0,0 +1,68 @@
+'''
+Helper functions for the ADC code. Davidson algorithm adapted from
+Coupled-Cluster/RHF/EOM_CCSD.py.
+'''
+
+__authors__ = 'Oliver J. Backhouse'
+__credits__ = ['T. Daniel Crawford', 'Andrew M. James', 'Oliver J. Backhouse']
+
+__copyright__ = '(c) 2014-2020, The Psi4NumPy Developers'
+__license__ = 'BSD-3-Clause'
+__date__ = '2018-03-01'
+
+import time
+import numpy as np
+
+
+def davidson(matrix, guesses, diag, maxiter=None, sort_via_abs=True, tol=1e-10, nvec_per_root=20):
```

Comment:
One could also use `scipy.sparse.linalg.eigsh` for this, requires less code 😉 

Details: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html

Reply:
I found that convergence is far slower with that, either because it's Lanczos or the lack of preconditioning? Not sure. Either way, even with this small example the EE runtime is pretty bad with `scipy.sparse.linalg.eigsh`.

Reply:
You're right, I thought `eigsh` supported preconditioning by specifying an extra `LinearOperator`, which does not seem to be the case...